src/main/java/com/google/googleidentity/oauth2/endpoint/TokenEndpoint.java

```diff
@@ -100,10 +118,29 @@ public OAuth2Request parseOAuth2RequestFromHttpRequest(HttpServletRequest reques
                 .setClientId(
                             OAuth2Utils.getClientSession(request).getClient().get().getClientId());
 
+        switch (grantType) {
+            case OAuth2Constants.GrantType.AUTHORIZATION_CODE:
+                oauth2RequestBuilder.getRequestBodyBuilder()
+                        .setGrantType(OAuth2Request.RequestBody.GrantType.AUTHORIZATION_CODE);
+                break;
+            case OAuth2Constants.GrantType.REFRESH_TOKEN:
+                oauth2RequestBuilder.getRequestBodyBuilder()
+                        .setGrantType(OAuth2Request.RequestBody.GrantType.REFRESH_TOKEN);
+                break;
+            case OAuth2Constants.GrantType.JWT_ASSERTION:
+                oauth2RequestBuilder.getRequestBodyBuilder()
+                        .setGrantType(OAuth2Request.RequestBody.GrantType.JWT_ASSERTION);
+                break;
+            default:
+                // Will never happen since we have validated it
+                break;
+        }
         oauth2RequestBuilder.getRequestBodyBuilder()
-                .setGrantType(grantType).setResponseType(OAuth2Constants.ResponseType.TOKEN);
+                .setResponseType(ResponseType.TOKEN);
 
         if (grantType.equals(OAuth2Constants.GrantType.AUTHORIZATION_CODE)) {
+            oauth2RequestBuilder.getRequestBodyBuilder()
+                    .setGrantType(OAuth2Request.RequestBody.GrantType.AUTHORIZATION_CODE);
```

Comment:
This can be removed. It has already been done above in the switch statement. Same with other cases below that set GrantType again.